platform.gcc.arch: support for AMD CPUs

volth opened the pull request:
support for AMD CPUs in ```localSystem.hostPlatform.platform.gcc.arch```

I have (or can rent) a test park of
 - [ ] ```AMD Opteron(tm) X2150 APU``` (corresponds to ```-march=btver2```)
 - [ ] ```AMD Opteron(tm) Processor 3365``` (corresponds to ```-march=bdver2```)
 - [ ] ```AMD EPYC 7401P 24-Core Processor``` (this one is problematic, ```-march=bdver4``` as well as ```-march=znver1``` enable too many features; the GCC flags should be either ```-march=bdver4 -mno-fma4 -mno-tbm -mno-xop -mno-lwp``` or ```-march=znver1 -mno-clzero mno-mwaitx -mno-xsaves```, which is impossible to set within constrains of #59225)
 - [x] ```AMD Ryzen 5 2400G with Radeon Vega Graphics``` (corresponds to ```-march=znver1```)

cc: @matthewbauer @7c6f434c

lopsided98 commented:
Isn't this a huge pain to maintain? Every time a new architecture is added, every package that uses SSE/AVX will need a new special case added, even if the architecture doesn't add any new features that are relevant to the package. How will you ensure that packages don't get missed when architectures are added in the future?

Right now I'm working on a package that uses SSE, and I need a special case for each architecture, even though all of the cases except for the default are the same.

It seems like this would be much more maintainable if there were functions that could be used to decide whether a particular platform supports a certain feature. This would only require changes in a single centralized place when a new architecture is added.

volth commented:
Yes, tables like [this](https://github.com/NixOS/nixpkgs/pull/59148/files#diff-cf9da652b842471fbbddf1128acc6085R270) are better to have somewhere in ```lib``` with a possibility to query for a particular CPU feature.

lopsided98 commented:
BTW, the package I am adding a CPU feature table to is `g2o`, in #61655.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/targeting-particular-x86-64-generation-haswell-ivybridge-skylake/2280/27

Ericson2314 commented:
This looks good to me, sorry I didn't see it earlier.

Ericson2314 commented:
CC @matthewbauer

Ericson2314 commented:
And yes I would like to see the tables move into `lib` eventually too.

FRidh commented:
For lib, could we perhaps reuse https://github.com/archspec/archspec-json/blob/master/cpu/microarchitectures.json? The archspec package is broken out of the Spack package manager for HPC. 
https://discuss.python.org/t/archspec-a-library-for-labeling-optimized-binaries/3149

vorot93 commented:
Any chance this could land in master soon?

volth commented:
yes, it might need adjustments for gcc 8, 9, 10 and clang 9, 10, 11

volth commented:
> Yes, tables like [this](https://github.com/NixOS/nixpkgs/pull/59148/files#diff-cf9da652b842471fbbddf1128acc6085) are better to have somewhere in `lib` with a possibility to query for a particular CPU feature.

For simplicity, I plan only to add AMD architectures here next to Intel and left moving tables to `lib` for another PR

jonringer commented on pkgs/development/libraries/dlib/default.nix:
```diff
@@ -2,7 +2,7 @@
 
   # see http://dlib.net/compile.html
-, avxSupport ? true
+, avxSupport ? builtins.elem (stdenv.hostPlatform.platform.gcc.arch or "default") [ "sandybridge" "ivybridge" "haswell" "broadwell" "skylake" "skylake-avx512" "btver2" "bdver1" "bdver2" "bdver3" "bdver4" "znver1"]
```
I feel like this could be nicer to work with if it was something like `stdenv.hostPlatform.platform.gcc.arch.hasAvxSupport`, where gcc knows whether avx is supported

conferno replied:
`stdenv.hostPlatform.platform` is a plain attrset, without any functions, so I added the functions into `stdenv.hostPlatform`, next to `.isx86_64` and other predicates

volth commented on pkgs/build-support/cc-wrapper/default.nix:
```diff
@@ -72,6 +72,7 @@ let
         knm            = versionAtLeast ccVersion "8.0";
         # AMD
         znver1         = versionAtLeast ccVersion "6.0";
+        znver2         = versionAtLeast ccVersion "9.0";
```
I am not sure when `znver2` support has been added no `gcc` and `clang`, so these version numbers are not final yet

volth commented on pkgs/build-support/cc-wrapper/default.nix:
```diff
@@ -81,6 +82,7 @@ let
         knm            = versionAtLeast ccVersion "7.0";
         # AMD
         znver1         = versionAtLeast ccVersion "4.0";
+        znver2         = versionAtLeast ccVersion "8.0";
```
I am not sure when `znver2` support has been added no `gcc` and `clang`, so these version numbers are not final yet

volth commented on lib/systems/architectures.nix:
```diff
@@ -0,0 +1,59 @@
+    broadwell      = [ "westmere" "sandybridge" "ivybridge" "haswell" ];
+    skylake        = [ "westmere" "sandybridge" "ivybridge" "haswell" "broadwell" ];
+    skylake-avx512 = [ "westmere" "sandybridge" "ivybridge" "haswell" "broadwell" "skylake" ];
+    # AMD
```
For AMD it is not so linear, some features have been removed over time (for example `fma4`), so be careful filling this table

volth replied:
The empty page is the safest :)

volth commented:
> Any chance this could land in master soon? I have a `znver1` machine at hand to help with testing.

@vorot93
Please cherry-pick the PR to your copy `master` to test.

I rebased it, added AMD CPUs (including `znver2` (Zen 2) which was not yet available) and moved CPU tables under `lib/` as many people suggested here.

It should be ready to test

jonringer commented:
@volth if you need additional testing, I have a 3990X, and a 3990X at home.

volth commented:
> @volth if you need additional testing, I have a 3990X, and a 3990X at home.

That would be nice.

I think we need to end up with a documentation page like https://wiki.gentoo.org/wiki/Ryzen after collection testing results.
It is especially important for AMD whose processors do not correspond to `-march=` values exactly (like EPYC 7401P case in the first message).

Ericson2314 commented on lib/systems/architectures.nix:
```diff
@@ -0,0 +1,75 @@
+    # x86_64 Intel
+    default        = [ ];
+    westmere       = [ ];
+    sandybridge    = [ "westmere" ];
```
How about
```suggestion
    sandybridge    = [ "westmere" ] ++ inferiors.westmere;
```
and so forth?

Ericson2314 commented:
Should we integrate this with `lib/systems/parse.nix` more? Is not `sandybridge-unknown-linux-gnu` a valid triple, for example?

Ericson2314 commented on lib/systems/architectures.nix:
```diff
@@ -0,0 +1,76 @@
+  };
+
+  predicates = rec {
+    featureSupport = feature: x: builtins.elem feature features.${x};
```
This should be let-bound because the `mapAttrs` below will apply the outer lambda to the wrong arg.

Ericson2314 commented:
> Hm, I do not know. Does the 1st component of the triple always equal to `-march=` value ?

No, but when it does (e.g. `armv7-m`) we track both. See https://github.com/NixOS/nixpkgs/blob/70cfd9d25e1d4f5a40f5d0a518f0749635792667/lib/systems/parse.nix#L73

> Also, gccarch might be extended from the -march= value to the list of gcc flags (to cover weird CPUs like AMD EPYC 7401P 24-Core Processor described above) and that would not fit into a single word in the triple

We should still distinguish the `march` flag, but sure we can have some "extra flags" needed to fill the difference.

See 
https://github.com/NixOS/nixpkgs/blob/70cfd9d25e1d4f5a40f5d0a518f0749635792667/pkgs/development/compilers/gcc/common/platform-flags.nix

We should use `targetPlatform.parsed.arch`, if it exists, to be another way to provide the `march`, and then can also provide the extra flags however you like.

Ericson2314 commented:
@volth I don't want to hold this up indefiniely / let the perfect be the enemy of the good, but based on my last comment did you take a look at integrating with parse.nix some more? I just want to reign in complexity wherever I can, and also reign in the potential for arch-related info to drift out of sync.

jonringer commented:
Is there a command I can run to verify my 3990X and 3900X are to match the extensions in builds?

volth commented:
I am not sure I understood the spirit of the triples and how far the
changes could go (where "sandybridge" of
"sandybridge-unknown-linux-gnu" would come from? Will the be
"sandybridge-linux" system along with "x86_64-linux"? How would it
affect meta.platforms? etc. ). It would be faster and easier if you
make the changes here or in a new pr

On 9/2/20, John Ericson <notifications@github.com> wrote:
> @volth I don't want to hold this up indefiniely / let the perfect be the
> enemy of the good, but based on my last comment did you take a look at
> integrating with parse.nix some more? I just want to reign in complexity
> wherever I can, and also reign in the potential for arch-related info to
> drift out of sync.
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/NixOS/nixpkgs/pull/61019#issuecomment-685155975

Ericson2314 commented:
Fair enough.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/tensorflow-slower-as-nixos-native-than-inside-a-docker-container/9967/4
